<a href="https://colab.research.google.com/github/monte137/ImplementacionSinLibrerias/blob/main/ImplementacionSinLibrerias.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Se utilizará una base de datos de pacientes que han sufrido un derrame cerebral y otros que no. La finalidad es clasificar las caracteristicas para predecir correctamente si el paciente sufrio un derrame o no

In [4]:
import pandas as pd
import numpy as np
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [6]:
data= pd.read_csv('/content/drive/MyDrive/ConcentracionCienciaDeDatos/Modulo2/full_data.csv',index_col=False)
data.head()

gender   age  hypertension  heart_disease ever_married      work_type  \
0    Male  67.0             0              1          Yes        Private   
1    Male  80.0             0              1          Yes        Private   
2  Female  49.0             0              0          Yes        Private   
3  Female  79.0             1              0          Yes  Self-employed   
4    Male  81.0             0              0          Yes        Private   

  Residence_type  avg_glucose_level   bmi   smoking_status  stroke  
0          Urban             228.69  36.6  formerly smoked       1  
1          Rural             105.92  32.5     never smoked       1  
2          Urban             171.23  34.4           smokes       1  
3          Rural             174.12  24.0     never smoked       1  
4          Urban             186.21  29.0  formerly smoked       1

Modelo

In [12]:
clean= {'gender': {'Male': 0,'Female':1},'ever_married':{'Yes':1,'No':0},'Residence_type':{'Urban':1,'Rural':0},'smoking_status':{'smokes':2,'formerly smoked':1,'never smoked':0,'Unknown':0}}
data= data.replace(clean)

In [13]:
dx= data[['hypertension','heart_disease','smoking_status']].values
dy= data['stroke'].values
dx=dx[:300]
dy=dy[:300]

Regresion logistica, utilizando notacion matricial para la tecnica de gradiente descendiente.
W=[w1,w2,...,wm] 1xm \\
El vector w es el contiene nuestros pesos para cada x, en clase visto como los pesos θ. \\
B= cte \\
Visto en clase como θ cero \\
X= [Matriz] nxm \\
Y= [Vector] nx1 \
Considerando nuestra funcion log-loss vista en clase, los terminos para el gradiente descendiente quedarian:

$W= W - α dW^T$ donde: \\
$dW= (A-Y) X^T$ \\
$A= σ(W^T *X +B)$\\

$B= B - α dB$


In [14]:
def sigmoide(X):
  return 1/(1+np.exp(-X))

ite= 100000
alpha=.008

dx=dx.T
dy= dy.reshape(1,dx.shape[1])
m=dx.shape[1]
n=dx.shape[0]

W= np.zeros((n,1))
B= 0
for i in range(ite):
  S= np.dot(W.T, dx ) + B
  A= sigmoide(S)

  logloss= -(1/m)*np.sum(dy* np.log(A) + (1-dy) *np.log(1-A))

  dW= (1/m)*np.dot(A-dy,dx.T)
  dB= (1/m)*np.sum(A-dy)

  W= W-alpha*dW.T 
  B= B-alpha*dB

  if (i%(ite/10)==0):
    print('Costo despues de',i, 'iteraciones, es:',logloss)


Costo despues de 0 iteraciones, es: 0.6931471805599453
Costo despues de 10000 iteraciones, es: 0.5709928027810194
Costo despues de 20000 iteraciones, es: 0.569641436777377
Costo despues de 30000 iteraciones, es: 0.569530574665546
Costo despues de 40000 iteraciones, es: 0.5695189930758259
Costo despues de 50000 iteraciones, es: 0.569517664803008
Costo despues de 60000 iteraciones, es: 0.5695175071241038
Costo despues de 70000 iteraciones, es: 0.5695174881766706
Costo despues de 80000 iteraciones, es: 0.5695174858902491
Costo despues de 90000 iteraciones, es: 0.5695174856139442


Calculo de exactitud del modelo.

In [16]:
Z= np.dot(W.T,dx) + B
A= sigmoide(Z)

A= A>.5

A= np.array(A,dtype= 'int64')

prec= (1 - np.sum(np.absolute(A-dy))/dy.shape[1])
prec

0.6933333333333334

Se obtiene un exactitud media de .69333, la cual considerando la implementación podría mejorar cambiando el método de optimización utilizado.